### PR TITLE
Fix anchor links in Wiki

### LINKED ISSUE #558: wiki: links to anchors don't work
neither inside one page nor from one page to another

### PR DESCRIPTION
**Pull request by JuliusR**
Fixes #558.

### CODE REVIEW

**Comment by JuliusR** [comment]
I am not sure if this works in all cases. We should check it carefully before merging.

File: plugins/wiki/lib/foodsoft_wiki/wiki_parser.rb [hunk 1]
```diff
@@ -9,10 +9,13 @@ class WikiParser < WikiCloth::Parser
...
     end
 
-    link_attributes_for do |page|
+    link_attributes_for do |page_anchor|
+      page, anchor = page_anchor.split(/#/, 2)
       permalink = Page.permalink(page)
-      if Page.exists?(:permalink => permalink)
-        { href: url_for(:wiki_page_path, permalink: permalink) }
+      if page.empty?
+        { href: '#' + anchor }
```
Comment (wvengen) [inline plugins/wiki/lib/foodsoft_wiki/wiki_parser.rb]: what is `anchor` is `nil`?

Comment (JuliusR) [reply plugins/wiki/lib/foodsoft_wiki/wiki_parser.rb]: I did not find markup that produces that state.

I should note that markup such as 

```
[[|Custom label]]
```

still produces the following error message:

> translation missing: de.wikicloth_exception

This pull request is made to fix at least the typical use cases of anchors.

**Comment by JuliusR** [comment]
Here are some tests I can come up with:

```
== Existing page ==

== DEFAULT label ==

* [[Home]]
* [[Home#Baz]]
* [[Home#]]

* [[#Baz]]
* [[#]]

== CUSTOM label ==

* [[Home|Custom label Home]]
* [[Home#Baz|Custom label Home#Baz]]
* [[Home#|Custom label Home#]]

* [[#Baz|Custom label #Baz]]
* [[#|Custom label #]]

== Missing page ==

=== DEFAULT label ===

* [[Missing]]
* [[Missing#Baz]]
* [[Missing#]]

=== CUSTOM label ===

* [[Missing|Custom label Missing]]
* [[Missing#Baz|Custom label Missing#Baz]]
* [[Missing#|Custom label Missing#]]

== Invalid Markup ==

<nowiki>* [[|Custom label "space"]]</nowiki>

yields "translation missing: de.wikicloth_exception"

== Baz ==
```

And here is the html output:

```
<h2>
  <span class="mw-headline" id="Existing_page">Existing page</span>
</h2>
<h2>
  <a name="DEFAULT_label"></a><span class="mw-headline" id="DEFAULT_label">DEFAULT label</span>
</h2>
<ul>
  <li>
    <a href="/f/wiki/Home">Home</a>
  </li>
  <li>
    <a href="/f/wiki/Home#Baz">Home#Baz</a>
  </li>
  <li>
    <a href="/f/wiki/Home#">Home#</a>
  </li>
</ul>
<ul>
  <li>
    <a href="#Baz">#Baz</a>
  </li>
  <li>
    <a href="#">#</a>
  </li>
</ul>
<h2>
  <a name="CUSTOM_label"></a><span class="mw-headline" id="CUSTOM_label">CUSTOM label</span>
</h2>
<ul>
  <li>
    <a href="/f/wiki/Home">Custom label Home</a>
  </li>
  <li>
    <a href="/f/wiki/Home#Baz">Custom label Home#Baz</a>
  </li>
  <li>
    <a href="/f/wiki/Home#">Custom label Home#</a>
  </li>
</ul>
<ul>
  <li>
    <a href="#Baz">Custom label #Baz</a>
  </li>
  <li>
    <a href="#">Custom label #</a>
  </li>
</ul>
<h2>
  <a name="Missing_page"></a><span class="mw-headline" id="Missing_page">Missing page</span>
</h2>
<h3>
  <a name="DEFAULT_label-2"></a><span class="mw-headline" id="DEFAULT_label-2">DEFAULT label</span>
</h3>
<ul>
  <li>
    <a href="/f/pages/new?parent=Home&amp;title=Missing" class="new_wiki_link">Missing</a>
  </li>
  <li>
    <a href="/f/pages/new?parent=Home&amp;title=Missing" class="new_wiki_link">Missing#Baz</a>
  </li>
  <li>
    <a href="/f/pages/new?parent=Home&amp;title=Missing" class="new_wiki_link">Missing#</a>
  </li>
</ul>
<h3>
  <a name="CUSTOM_label-2"></a><span class="mw-headline" id="CUSTOM_label-2">CUSTOM label</span>
</h3>
<ul>
  <li>
    <a href="/f/pages/new?parent=Home&amp;title=Missing" class="new_wiki_link">Custom label Missing</a>
  </li>
  <li>
    <a href="/f/pages/new?parent=Home&amp;title=Missing" class="new_wiki_link">Custom label Missing#Baz</a>
  </li>
  <li>
    <a href="/f/pages/new?parent=Home&amp;title=Missing" class="new_wiki_link">Custom label Missing#</a>
  </li>
</ul>
<h2>
  <a name="Invalid_Markup"></a><span class="mw-headline" id="Invalid_Markup">Invalid Markup</span>
</h2>
<p>* [[|Custom label "space"]]</p>
<p>yields "translation missing: de.wikicloth_exception"</p>
<h2>
  <a name="Baz"></a><span class="mw-headline" id="Baz">Baz</span>
</h2>
```

**Comment by JuliusR** [comment]
How should we continue with this?

**Comment by JuliusR** [comment]
I forgot to consider another problematic case: Wiki pages are allowed to have `#` inside the title. That is bad because this pull requests breaks links such as
```
[[Wiki page with # in the title]]

[[Wiki page with # in the title#How_to_reference_this_anchor]]
```

**Comment by JuliusR** [comment]
I tested for the following wiki markup:

```
[[Existent]]

[[Existent with # in title]]

[[Existent with # and # in title]]



[[Existent#Anchor]]

[[Existent with # in title#Anchor]]

[[Existent with # and # in title#Anchor]]



[[Nonexistent]]

[[Nonexistent with # in title]]

[[Nonexistent with # and # in title]]



[[Nonexistent#Anchor]]

[[Nonexistent with # in title#Anchor]]

[[Nonexistent with # and # in title#Anchor]]



[[#Existent and confusing]]

[[#Nonexistent]]

[[#]]
```

Here is the output:
```
<p><a href="/f/wiki/Existent">Existent</a></p>
<p><a href="/f/wiki/Existent_with_%23_in_title">Existent with # in title</a></p>
<p><a href="/f/wiki/Existent_with_%23_and_%23_in_title">Existent with # and # in title</a></p>

<p><a href="/f/wiki/Existent#Anchor">Existent#Anchor</a></p>
<p><a href="/f/wiki/Existent_with_%23_in_title#Anchor">Existent with # in title#Anchor</a></p>
<p><a href="/f/wiki/Existent_with_%23_and_%23_in_title#Anchor">Existent with # and # in title#Anchor</a></p>

<p><a href="/f/pages/new?parent=Test&amp;title=Nonexistent" class="new_wiki_link">Nonexistent</a></p>
<p><a href="/f/pages/new?parent=Test&amp;title=Nonexistent+with+" class="new_wiki_link">Nonexistent with # in title</a></p>
<p><a href="/f/pages/new?parent=Test&amp;title=Nonexistent+with+" class="new_wiki_link">Nonexistent with # and # in title</a></p>

<p><a href="/f/pages/new?parent=Test&amp;title=Nonexistent" class="new_wiki_link">Nonexistent#Anchor</a></p>
<p><a href="/f/pages/new?parent=Test&amp;title=Nonexistent+with+" class="new_wiki_link">Nonexistent with # in title#Anchor</a></p>
<p><a href="/f/pages/new?parent=Test&amp;title=Nonexistent+with+" class="new_wiki_link">Nonexistent with # and # in title#Anchor</a></p>

<p><a href="/f/wiki/%23Existent_and_confusing">#Existent and confusing</a></p>
<p><a href="#Nonexistent">#Nonexistent</a></p>
<p><a href="#">#</a></p>
```

**Comment by JuliusR** [comment]
It is not perfect, but I can not think of a better fix for the number-sign problem.

If users do weird things (such as creating wiki pages with number signs in the title) then the outcome may be surprising. Compared to the current situation (no internal anchor links are currently possible), this is still an improvement (IMHO).

I can accept the compromise of this pull request now. What is your opinion?

File: plugins/wiki/lib/foodsoft_wiki/wiki_parser.rb (hunk 1 above)
Comment (JuliusR) [reply plugins/wiki/lib/foodsoft_wiki/wiki_parser.rb]: With the latest commit here (b1fabe8), `anchor` is obviously non-`nil`, right?

**Comment by wvengen** [comment]
Looks good. I'd use "hash" or "hash mark" instead of "number sign", but that may be a bikeshed issue.

**Comment by JuliusR** [comment]
Thanks for your second-round feedback. I do care about these details.

I just checked. The term "number sign" [is used](https://html.spec.whatwg.org/#syntax-references) in the HTML standard. According to the standard, "anchor" should be replaced by "hash-name reference". I have never read that term before, though.

In summary, I consider this ready to be merged.

**Pull request merged**